FNML/overview.md

```diff
@@ -33,33 +57,34 @@ The execution of such function would be described as follows:
     grel:stringOutput "THIS IS AN INPUT STRING." .
 ```
 
-To connect this function with the RML mapping document, we make use of a `fnml:FunctionMap`
+To connect this function with the RML mapping document, we make use of a `fnml:Execution`
 
 ```turtle "example": "using toUppercase in an RML mapping"
 <#Person_Mapping>
-    rml:logicalSource <#LogicalSource> ;                  # Specify the data source
-    rr:subjectMap <#SubjectMap> ;                         # Specify the subject
-    rr:predicateObjectMap <#NameMapping> .                # Specify the predicate-object-map
+    rml:logicalSource <#LogicalSource> ;       # Specify the data source
+    rr:subjectMap <#SubjectMap> ;              # Specify the subject
+    rr:predicateObjectMap <#NameMapping> .     # Specify the predicate-object-map
 
 <#NameMapping>
-    rr:predicate dbo:title ;                              # Specify the predicate
-    rr:objectMap <#FunctionMap> .                         # Specify the object-map
-
-<#FunctionMap>
-    fnml:functionValue [                                  # The object is the result of the function
-        rr:predicateObjectMap [
-            rr:predicate fno:executes ;                   # Execute the function&hellip;
-            rr:objectMap [ rr:constant grel:toUppercase ] # grel:toUppercase
-        ] ;
-        rr:predicateObjectMap [
-            rr:predicate grel:inputString ;
-            rr:objectMap [ rr:reference "name" ]          # Use as input the "name" reference
-        ]
-    ] .
+    rr:predicate dbo:title ;                   # Specify the predicate
+    rr:objectMap <#ExecutionTermMap> .         # Specify the object-map
+
+<#ExecutionTermMap> a fnml:ExecutionTermMap ;  # A Term Map: the link between RML and an FnO Execution
+    fnml:execution <#Execution> ;              # Link to the execution
+    fnml:output grel:stringOutput .            # Specify which output of the execution is taken
+
+<#Execution> a fnml:Execution ;                # A new class, unrelated to existing RML constructs
+    fnml:function grel:toUppercase ;           # Specify which FnO function
+    fnml:inputParameters                       # Specify the input parameters
+        [
+            fnml:predicate grel:stringInput ;  # Specify this specific parameter
+            rr:reference "name" .              # Specify the value using existing Term Map constructions
+        ] .
 ```
 
-Before the `name`-value is referenced,
-the value is first used as `grel:inputString`-parameter
+The `name`-value is not referenced directly,
+instead, its value is used as `grel:inputString`-parameter
 for the `grel:toUppercase`-function.
-The output of that function is then used as object
-within the `<#NameMapping>`
+After execution, the `grel:stringOutput` result of that function is returned to generate the object
+within the `<#NameMapping>`.
+We make use of an intermediate `<#ExecutionTermMap>` so that we can reuse the output of an execution in multiple TermMaps.
```

Comment:
both: if an execution returns multiple outputs (eg, a result and a status code), by referring to the same execution, you can use both outputs in different locations of the same mapping. If you don't reuse, you can't specify the difference between 'using 2 outputs from one execution' vs 'use a different output from 2 executions'. See example below for both cases (the latter one quite artificial, I know)

```turtle
ex:parseNameToFirstAndLastName a fno:Function ;
    fno:name "parses a name into first name and last name, e.g. 'Ben De Meester' becomes {firstname: 'Ben', lastname: 'De Meester'}" ;
    fno:expects ( [ fno:predicate grel:stringInput ] ) ;
    fno:output ( [ fno:predicate grel:firstnameOutput ] [ fno:predicate grel:firstnameOutput ] ) .

<#Person_Mapping>
    rml:logicalSource <#LogicalSource> ;
    rr:subjectMap <#SubjectMap> ;
    rr:predicateObjectMap <#FirstNameMapping> , <#LastNameMapping>, <#RandomWordMapping1> , <#RandomWordMapping2> ,

<#FirstNameMapping>
    rr:predicate foaf:firstname ;
    rr:objectMap <#ExecutionTermMapFirst> .

<#LastNameMapping>
    rr:predicate foaf:lastname ;
    rr:objectMap <#ExecutionTermMapLast> .

<#RandomWordMapping1>
    rr:predicate ex:randomWord1 ;
    rr:objectMap <#ExecutionRandom1> .

<#RandomWordMapping2>
    rr:predicate ex:randomWord2 ;
    rr:objectMap <#ExecutionRandom2> .
    

<#ExecutionTermMapFirst> a fnml:ExecutionTermMap ;
    fnml:execution <#Execution> ;                    # By referring to the same execution, I can specify to reuse the result of one execution
    fnml:output ex:firstnameOutput .

<#ExecutionTermMapFirst> a fnml:ExecutionTermMap ;
    fnml:execution <#Execution> ;
    fnml:output ex:lastnameOutput .

<#Execution> a fnml:Execution ;
    fnml:function ex:parseNameToFirstAndLastName ;
    fnml:inputParameters
        [
            a fnml:ParameterTermMap ;
            fnml:predicate grel:stringInput ;
            rr:reference "name" .
        ] .

<#ExecutionRandom1> a fnml:ExecutionTermMap ;
    fnml:execution <#ExecutionRandom1> ;                    # Without an Execution, I cannot specify I want to execute `ex:take2RandomWords` twice
    fnml:output ex:out1 .

<#ExecutionRandom2> a fnml:ExecutionTermMap ;
    fnml:execution <#ExecutionRandom2> ;
    fnml:output ex:out2 .

<#ExecutionRandom1> a fnml:Execution ;
    fnml:function ex:take2RandomWords ;
    fnml:inputParameters
        [
            a fnml:ParameterTermMap ;
            fnml:predicate grel:stringInput ;
            rr:reference "name" .
        ] .

<#ExecutionRandom2> a fnml:Execution ;
    fnml:function ex:take2RandomWords ;
    fnml:inputParameters
        [
            a fnml:ParameterTermMap ;
            fnml:predicate grel:stringInput ;
            rr:reference "name" .
        ] .
```

Reply:
Ah I see.
Hmm, but so in this approach we would need to use IRI of the Execution(Map) to identify a specific execution. I don't think the name (IRI / blanknode id) of a mapping construct is used liked this anywhere else in the language. I also don't think we should assign meaning to the naming of mapping constructs.

Reply:
Could we support the 'use a different output from 2 executions' in another way? Or should we not overanalyze?

> Note: I made a typo in the example, I now clarified the difference between eg `ExecutionTermRandom1` and `ExecutionRandom1`

Reply:
By generating a subject IRI / bnode for the execution :grimacing:?

Reply:
😂 Back to square one! :D no but how would that work exactly for the case where you want to reuse an execution across term maps? Do you mean like below?

```turtle
<#Person_Mapping>
    rml:logicalSource <#LogicalSource> ;
    rr:subjectMap <#SubjectMap> ;
    rr:predicateObjectMap <#FirstNameMapping> , <#LastNameMapping> .

<#FirstNameMapping>
    rr:predicate foaf:firstname ;
    rr:objectMap <#ExecutionFirst> .

<#LastNameMapping>
    rr:predicate foaf:lastname ;
    rr:objectMap <#ExecutionLast> .  

<#ExecutionFirst>
    # let's assume we can do this using TriplesMap-like constructs, and work out the details later
    rr:subjectMap [ rr:constant "ReusableExecution" ]
    fnml:function ex:parseNameToFirstAndLastName ;
    fnml:inputParameters
        [
            a fnml:ParameterTermMap ;
            fnml:predicate grel:stringInput ;
            rr:reference "name" .
        ] .
    fnml:output ex:firstnameOutput .

<#ExecutionLast>
    rr:subjectMap [ rr:constant "ReusableExecution" ] ; # because it generates the same subject, you can reuse the execution, so same parameters
    fnml:output ex:lastnameOutput .
```

Reply:
Well, I don't know if that would parse...

Something like this
```turtle
<#Person_Mapping>
    rml:logicalSource <#LogicalSource> ;
    rr:subjectMap <#SubjectMap> ;
    rr:predicateObjectMap <#FirstNameMapping> , <#LastNameMapping> .

<#FirstNameMapping>
    rr:predicate foaf:firstname ;
    rr:objectMap [
      fnml:executes <#ExecutionFirst> ; # Or <#ExecutionLast> doesn't matter because the result is the same
      fnml:output ex:firstnameOutput ;
    ] .

<#LastNameMapping>
    rr:predicate foaf:lastname ;
    rr:objectMap [
      fnml:executes <#ExecutionFirst> ;  # Or <#ExecutionLast> doesn't matter because the result is the same
      fnml:output ex:lastnameOutput ;
    ] .

<#ExecutionFirst>
    # let's assume we can do this using TriplesMap-like constructs, and work out the details later
    rr:subjectMap [ rr:constant "ReusableExecution" ]
    fnml:function ex:parseNameToFirstAndLastName ;
    fnml:inputParameters
        [
            a fnml:ParameterTermMap ;
            fnml:predicate grel:stringInput ;
            rr:reference "name" .
        ] .

<#ExecutionLast>
    rr:subjectMap [ rr:constant "ReusableExecution" ] ; # because it generates the same subject, you can reuse the execution, so same parameters
    fnml:function ex:parseNameToFirstAndLastName ;
    fnml:inputParameters
        [
            a fnml:ParameterTermMap ;
            fnml:predicate grel:stringInput ;
            rr:reference "name" .
        ] .
```

Reply:
Hmmm, I'm confused, how is this different from my initial clarification at https://github.com/kg-construct/rml-fno-spec/pull/45#discussion_r816869238 ? You also refer to the same `<#ExecutionFirst>` via an intermediate term map? Are we agreeing? 😅

Reply:
The difference is that in your initial clarification `<#Execution>` has no subject map, but you seem to imply that the IRI / BNode id of the mapping construct `<#Execution>` should be used to indicate that the function execution is the same.

So, if you were to generate FNO execution triples, how do you determine the subject of those triples?

Another concern with such an approach is that it impedes generic RDF to Object mapping approaches where information such as node IRIs/BNode ids might get lost.

Reply:
> Are we agreeing? 😅

Almost 😄

Reply:
> The difference is that in your initial clarification `<#Execution>` has no subject map, but you seem to imply that the IRI / BNode id of the mapping construct `<#Execution>` should be used to indicate that the function execution is the same.
> 
> So, if you were to generate FNO execution triples, how do you determine the subject of those triples?

Ah yeah, didn't come to that part yet, you can probably extend to that

> 
> Another concern with such an approach is that it impedes generic RDF to Object mapping approaches where information such as node IRIs/BNode ids might get lost.

You mean, in the code of an RML engine software? I have the feeling that an approach that loses node IRIs is a bad approach, but I don't understand fully what you're saying here

I'm not sure I understand the difference between 'You use the actual IRI of the execution to differentiate between executions' and 'You use the subjectmap of the execution to differentiate between executions' (except that you can reuse the latter for the execution triples, and that your RML can be prettier bc you can use blank nodes for executions)

So: I'm totally ok with a construct that adds a subject(map) to an execution :D

If that was the last hurdle, we are now in Official Agreement™

Reply:
> > The difference is that in your initial clarification `<#Execution>` has no subject map, but you seem to imply that the IRI / BNode id of the mapping construct `<#Execution>` should be used to indicate that the function execution is the same.
> > So, if you were to generate FNO execution triples, how do you determine the subject of those triples?
> 
> Ah yeah, didn't come to that part yet, you can probably extend to that
> 
> > Another concern with such an approach is that it impedes generic RDF to Object mapping approaches where information such as node IRIs/BNode ids might get lost.
> 
> You mean, in the code of an RML engine software? I have the feeling that an approach that loses node IRIs is a bad approach, but I don't understand fully what you're saying here

Yep analogous to mapping from JSON/YAML to objects. e.g. how would this work in YARRRML?

> 
> I'm not sure I understand the difference between 'You use the actual IRI of the execution to differentiate between executions' and 'You use the subjectmap of the execution to differentiate between executions' (except that you can reuse the latter for the execution triples, and that your RML can be prettier bc you can use blank nodes for executions)

One is part of the mapping structure, and the other is expressed using the mapping structure.
Also i feel it makes more sense to use the IRI of the FNO execution conceptually.

> 
> So: I'm totally ok with a construct that adds a subject(map) to an execution :D
> 
> If that was the last hurdle, we are now in Official Agreement™

OK 😄

Reply:
> > > Another concern with such an approach is that it impedes generic RDF to Object mapping approaches where information such as node IRIs/BNode ids might get lost.
> > 
> > 
> > You mean, in the code of an RML engine software? I have the feeling that an approach that loses node IRIs is a bad approach, but I don't understand fully what you're saying here
> 
> Yep analogous to mapping from JSON/YAML to objects. e.g. how would this work in YARRRML?
> 

I would say similar to how logicalsources work in YARRRML: you can either include it inline (similar to a blank node), or refer to a key that is part of top-level construct (similar to a named node). That allows to make that overhead only necessary in cases it would be needed (see examples 11 and 15 at https://rml.io/yarrrml/spec/#examples)

> One is part of the mapping structure, and the other is expressed using the mapping structure. Also i feel it makes more sense to use the IRI of the FNO execution conceptually.

I agree on the conceptual idea of being able to specify the FnO Execution identifier. But considering your first part: isn't this similar to referring to a parenttriplesmap? I think spec-wise, it should be possible to refer to a parenttriplesmap fully using blanknodes, what's the difference with that and functions?

```turtle
<#DeptTriplesMap>
    rr:logicalTable [ rr:tableName "DEPT" ];
    rr:subjectMap [
        rr:template "department/{DEPTNO}";
        rr:class ex:Department;
    ];
    rr:predicateObjectMap [
        rr:predicate ex:location;
        rr:objectMap [ rr:parentTriplesMap <#SiteTriplesMap> ];
    ].

<#SiteTriplesMap>
    rr:logicalTable [ rr:tableName "DEPT" ];
    rr:subjectMap [
        rr:template "site/{LOC}";
        rr:class ex:Site;
    ];
    rr:predicateObjectMap [
        rr:predicate ex:siteName;
        rr:objectMap [ ex:column "LOC" ];
    ].
```

should be equivalent to

```turtle
<#DeptTriplesMap>
    rr:logicalTable [ rr:tableName "DEPT" ];
    rr:subjectMap [
        rr:template "department/{DEPTNO}";
        rr:class ex:Department;
    ];
    rr:predicateObjectMap [
        rr:predicate ex:location;
        rr:objectMap [ rr:parentTriplesMap [
            rr:logicalTable [ rr:tableName "DEPT" ];
            rr:subjectMap [
                rr:template "site/{LOC}";
                rr:class ex:Site;
            ];
            rr:predicateObjectMap [
                rr:predicate ex:siteName;
                rr:objectMap [ ex:column "LOC" ];
            ].
        ] ];
    ].
```

Reply:
Right, but this is an example of inlining and reusing a mapping construct, not the result of a(n intermediate) mapping (which in this case produces the fno triples), which is subtly different. Or am I missing something?

Reply:
Yeah, it depends on whether you see `fnml:Execution` as a special (new) mapping construct or a type of triplesmap. I'm fine with either option though :D